{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import doekbase.data_api\n",
    "from doekbase.data_api.annotation.genome_annotation.api import GenomeAnnotationAPI\n",
    "import time\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_gff(gene,genome_annotation, taxon_id, source_info):\n",
    "    \n",
    "    print(gene)\n",
    "    #retrieve pairwise inter-feature relationships between genes, mRNAs, CDSs\n",
    "    #the below section of code has method timings torecord performance  \n",
    "    t0 = time.time()\n",
    "    listmrna=genome_annotation.get_mrna_by_gene([gene])[gene]    \n",
    "    t1 = time.time()\n",
    "    print \"get_mrna_by_gene \" , (t1-t0), \"s\"\n",
    "    #print(listmrna)\n",
    "    \n",
    "    #t0 = time.time()\n",
    "    #listcds=genome_annotation.get_cds_by_gene([gene])[gene] \n",
    "    #t1 = time.time()\n",
    "    #print \"get_cds_by_gene \" , (t1-t0), \"s\"   \n",
    "    #print(listcds)\n",
    "        \n",
    "    if len(listmrna) > 0:\n",
    "        t0 = time.time()\n",
    "        print(listmrna)\n",
    "        mrna_cds=genome_annotation.get_cds_by_mrna(listmrna)\n",
    "        t1 = time.time()\n",
    "        print \"get_cds_by_mrna \" , (t1-t0), \"s\"\n",
    "        #print(mrna_cds)\n",
    "    \n",
    "        #t0 = time.time()\n",
    "        #cds_mrna=genome_annotation.get_mrna_by_cds(listcds)\n",
    "        #t1 = time.time()\n",
    "        #print \"get_mrna_by_cds \" , (t1-t0), \"s\"\n",
    "\n",
    "        t0 = time.time()\n",
    "        get_mrna_exons = genome_annotation.get_mrna_exons(listmrna)\n",
    "        t1 = time.time()\n",
    "        print \"get_mrna_exons \" , (t1-t0), \"s\"\n",
    "        #print get_mrna_exons\n",
    "        \n",
    "        #retrieve locations for mRNAs\n",
    "        t0 = time.time()\n",
    "        mrna_location=genome_annotation.get_feature_locations(listmrna) \n",
    "        t1 = time.time()\n",
    "        print \"get_feature_locations \" , (t1-t0), \"s\"\n",
    "\n",
    "\n",
    "    #retrieve locations for genes, mRNAs, and cds\n",
    "    t0 = time.time()\n",
    "    gene_location=genome_annotation.get_feature_locations([gene]) \n",
    "    t1 = time.time()\n",
    "    print \"get_feature_locations \" , (t1-t0), \"s\"\n",
    "    \n",
    "    #t0 = time.time()\n",
    "    #cds_location=genome_annotation.get_feature_locations(listcds) \n",
    "    #t1 = time.time()\n",
    "    #print \"get_feature_locations \" , (t1-t0), \"s\"\n",
    "    \n",
    "    string=\"\"\n",
    "    \n",
    "    string+=\"##sequence-region \"+gene_location[gene_location.keys()[0]][0]['contig_id']+\"\\t\"+\"start\"+\"\\t\"+\"stop\"+\"\\n\"##NC_003070.9 1 30427671\n",
    "    if taxon_id != -1:\n",
    "        string+=\"##species http://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=\"+str(taxon_id)+\"\\n\"#http://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=3702\n",
    "    else :\n",
    "        string+=\"##species unknown\\n\"\n",
    "    \n",
    "\n",
    "    contig_id = \"\"\n",
    "    #output the gene data\n",
    "    for x in gene_location:\n",
    "        #print(type(x))\n",
    "        #print(x)\n",
    "        info=gene_location[x][0]\n",
    "        #print(type(info))\n",
    "        #print info\n",
    "        stop=info['start'] + info['length']\n",
    "        \n",
    "        contig_id = info['contig_id']\n",
    "        \n",
    "        #print \"source_info\"\n",
    "        #print(type(source_info))\n",
    "        #print(source_info)\n",
    "        #output aliases for gene as dbxref?\n",
    "        string+=info['contig_id']  + \"\\t\" + source_info['external_source']+\"\\tgene\\t\"  + str(info['start']) + \"\\t\" + str(stop) + \"\\t.\\t\" +  info['strand']+ \"\\t.\\t\" + \"ID=\" + x + \";\" + \"Name=\"+ x +\"\\n\"\n",
    "    \n",
    "    print \"len(listmrna)\"    \n",
    "    print(len(listmrna))\n",
    "    if len(listmrna) > 0:    \n",
    "    #output the mRNA data\n",
    "        for mrna in listmrna:\n",
    "            print(\"mrna \"+mrna)\n",
    "            cds=mrna_cds[mrna]\n",
    "            infomrna=mrna_location[mrna][0]\n",
    "            #print(type(infomrna))\n",
    "            #print(infomrna)\n",
    "            #infocds=cds_location[mrna_cds[mrna]]\n",
    "            stop=infomrna['start'] + infomrna['length']            \n",
    "            ###output mrna data\n",
    "            string+=infomrna['contig_id']  + \"\\t\" +source_info['external_source']+\"\\tmRNA\\t\"  + str(infomrna['start'])+ \"\\t\" + str(stop) + \"\\t.\\t\" +  infomrna['strand']+ \"\\t.\\t\" + \"ID=\" + mrna + \";\" + \"Parent=\"+ gene+\"\\n\"\n",
    "            \n",
    "            #print 'get_mrna_exons.keys()'\n",
    "            #print get_mrna_exons.keys()\n",
    "            \n",
    "            ##TODO output UTRs\n",
    "            for mrna in get_mrna_exons.keys():\n",
    "                #print(type(get_mrna_exons[mrna]))\n",
    "                #print(get_mrna_exons[mrna])\n",
    "                #print \"get_mrna_exons[mrna][i]['exon_location']\"\n",
    "                #print type(get_mrna_exons[mrna][0]['exon_location'])\n",
    "                ###output exon data\n",
    "                for i in range(0, len(get_mrna_exons[mrna])):\n",
    "                    #print \"mrna, gene\"\n",
    "                    #print type(mrna)\n",
    "                    #print type(gene)\n",
    "                    stop= get_mrna_exons[mrna][i]['exon_location']['start'] + get_mrna_exons[mrna][i]['exon_location']['length']\n",
    "                    string+= get_mrna_exons[mrna][i]['exon_location']['contig_id']  + \"\\t\" +source_info['external_source'] + \"\\texon\\t\"  + str(get_mrna_exons[mrna][i]['exon_location']['start'])+ \"\\t\" + str(stop) + \"\\t.\\t\" +  get_mrna_exons[mrna][i]['exon_location']['strand']+ \"\\t.\\t\" + \"ID=\" + mrna +\".exon.\" + str(i)  + \";\" + \"Parent=\"+ mrna+\"\\n\"\n",
    "                ###output CDS data\n",
    "                for i in range(0, len(get_mrna_exons[mrna])):\n",
    "                    stop= get_mrna_exons[mrna][i]['exon_location']['start'] + get_mrna_exons[mrna][i]['exon_location']['length']\n",
    "                    ##***TODO: set PHASE for CDS\n",
    "                    string+= get_mrna_exons[mrna][i]['exon_location']['contig_id']  + \"\\t\" +source_info['external_source'] + \"\\tCDS\\t\"  + str(get_mrna_exons[mrna][i]['exon_location']['start'])+ \"\\t\" + str(stop) + \"\\t.\\t\" +  get_mrna_exons[mrna][i]['exon_location']['strand']+ \"\\t.\\t\" + \"ID=\" + mrna +\".CDS.\" + str(i)  + \";\" + \"Parent=\"+ mrna+\"\\n\"\n",
    "\n",
    "    return string       \n",
    "\n",
    "#hardcoded test with 5 random genes\n",
    "def runDemo(ws_url='https://ci.kbase.us/services/ws/', ws=\"ReferenceEnsemblPlantGenomeAnnotations\", genome_id=\"280699_Ensembl\"):\n",
    "    #an example KBase reference genome\n",
    "    genome_ref = \"ReferenceEnsemblPlantGenomeAnnotations\"+\"/\"+genome_id\n",
    "    #instantiate a new genome annotation API\n",
    "    genome_annotation = GenomeAnnotationAPI(services = {\"workspace_service_url\": ws_url}, token=os.environ.get('KB_AUTH_TOKEN'), ref=genome_ref)\n",
    "    taxon = genome_annotation.get_taxon()\n",
    "    taxonid = taxon.get_taxonomic_id()\n",
    "    \n",
    "    if taxonid == -1 :\n",
    "        taxonid = \"unknown\"\n",
    "    \n",
    "    assembly = genome_annotation.get_assembly()\n",
    "    contig_ids = assembly.get_contig_ids()\n",
    "    sourceinfo = assembly.get_external_source_info()\n",
    "    genes=['gene_1','gene_2','gene_5', 'gene_100', 'gene_1000']\n",
    "    #genes=['gene_1']\n",
    "    \n",
    "    gffdata=\"\"\n",
    "    gffdata+=\"##gff-version 3\\n\"\n",
    "    gffdata+=\"#!gff-spec-version 1.21\\n\"\n",
    "    gffdata+=\"#!processor KBase GFF3 downloader\\n\"\n",
    "    #!genome-build TAIR10\n",
    "    #!genome-build-accession NCBI_Assembly:GCF_000001735.3   \n",
    "    \n",
    "    for s in genes:\n",
    "        gffdata+=get_gff(s,genome_annotation,taxonid, sourceinfo)\n",
    "\n",
    "    gffdata+=\"###\"\n",
    "    outfile = genome_id+'.gff'\n",
    "    print outfile\n",
    "    with open(outfile, 'w') as f:\n",
    "        f.write(gffdata)\n",
    "    \n",
    "#create GFF for an entire GenomeAnnotation\n",
    "def runGenome(ws_url='https://ci.kbase.us/services/ws/', ws=\"ReferenceEnsemblPlantGenomeAnnotations\", genome_id=\"280699_Ensembl\"):\n",
    "    #an example KBase reference genome\n",
    "    genome_ref = \"ReferenceEnsemblPlantGenomeAnnotations\"+\"/\"+genome_id#kb|g.166819\"\n",
    "    #instantiate a new genome annotation API\n",
    "    genome_annotation = GenomeAnnotationAPI(services = {\"workspace_service_url\": ws_url}, token=os.environ.get('KB_AUTH_TOKEN'), ref=genome_ref)\n",
    "    taxon = genome_annotation.get_taxon()\n",
    "    taxonid = taxon.get_taxonomic_id()\n",
    "    \n",
    "    if taxonid == -1 :\n",
    "        taxonid = \"unknown\"\n",
    "    \n",
    "    assembly = genome_annotation.get_assembly()\n",
    "    contig_ids = assembly.get_contig_ids()\n",
    "    sourceinfo = assembly.get_external_source_info()\n",
    "    \n",
    "    gffdata=\"\"\n",
    "    gffdata+=\"##gff-version 3\\n\"\n",
    "    gffdata+=\"#!gff-spec-version 1.21\\n\"\n",
    "    gffdata+=\"#!processor KBase GFF3 downloader\\n\"\n",
    "    #!genome-build TAIR10\n",
    "    #!genome-build-accession NCBI_Assembly:GCF_000001735.3  \n",
    "    \n",
    "    start=1\n",
    "    stop=100000000\n",
    "    \n",
    "    feat_data = []\n",
    "    print \"contig_ids len \"+ str(len(contig_ids))\n",
    "    for c in contig_ids:\n",
    "        regions = []\n",
    "        contig_add = {\"contig_id\": c, \"start\": start, \"length\": stop-start, \"strand\": \"+\"}\n",
    "        contig_add2 = {\"contig_id\": c, \"start\": stop, \"length\": stop-start, \"strand\": \"-\"}\n",
    "        regions.append(contig_add)\n",
    "        regions.append(contig_add2)\n",
    "        feat_data = genome_annotation.get_feature_ids(filters={\"region_list\":regions,\"type_list\":[\"gene\"]})['by_type']['gene']\n",
    "        \n",
    "        print \"feat_data len \"+ str(len(feat_data))\n",
    "        #print feat_data\n",
    "        #for s in genes:\n",
    "        for f in feat_data:\n",
    "            print f\n",
    "            gffdata+=get_gff(f,genome_annotation,taxonid, sourceinfo)\n",
    "            \n",
    "        #break after first contig for now    \n",
    "        break;\n",
    "\n",
    "    gffdata+=\"###\"\n",
    "    outfile = genome_id+'.gff'\n",
    "    print outfile\n",
    "    with open(outfile, 'w') as f:\n",
    "        f.write(gffdata)\n",
    "        \n",
    "    feat_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_1\n",
      "get_mrna_by_gene  1.67064189911 s\n",
      "[u'mRNA_1']\n",
      "get_cds_by_mrna  1.50494503975 s\n",
      "get_mrna_exons  1.49544405937 s\n",
      "get_feature_locations  1.95886707306 s\n",
      "get_feature_locations  2.66089892387 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_1\n",
      "gene_2\n",
      "get_mrna_by_gene  1.5816628933 s\n",
      "[u'mRNA_2']\n",
      "get_cds_by_mrna  1.58651709557 s\n",
      "get_mrna_exons  1.58610486984 s\n",
      "get_feature_locations  1.94034600258 s\n",
      "get_feature_locations  1.88635587692 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_2\n",
      "gene_5\n",
      "get_mrna_by_gene  1.54612207413 s\n",
      "[u'mRNA_3']\n",
      "get_cds_by_mrna  1.53069210052 s\n",
      "get_mrna_exons  1.67014789581 s\n",
      "get_feature_locations  2.21215605736 s\n",
      "get_feature_locations  2.13090801239 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_3\n",
      "gene_100\n",
      "get_mrna_by_gene  1.62239289284 s\n",
      "[u'mRNA_63']\n",
      "get_cds_by_mrna  1.63559103012 s\n",
      "get_mrna_exons  1.58180403709 s\n",
      "get_feature_locations  1.98508691788 s\n",
      "get_feature_locations  2.41667795181 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_63\n",
      "gene_1000\n",
      "get_mrna_by_gene  1.99698400497 s\n",
      "[u'mRNA_932']\n",
      "get_cds_by_mrna  1.54204797745 s\n",
      "get_mrna_exons  1.55971193314 s\n",
      "get_feature_locations  2.11794495583 s\n",
      "get_feature_locations  2.13708090782 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_932\n",
      "280699_Ensembl.gff\n"
     ]
    }
   ],
   "source": [
    "genomeid=\"280699_Ensembl\"\n",
    "wsthis = \"ReferenceEnsemblPlantGenomeAnnotations\"\n",
    "run(ws=wsthis, genome_id=genomeid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "contig_ids len 22\n",
      "feat_data len 363\n",
      "gene_3990\n",
      "gene_3990\n",
      "get_mrna_by_gene  2.17358088493 s\n",
      "[u'mRNA_3889']\n",
      "get_cds_by_mrna  2.24047613144 s\n",
      "get_mrna_exons  2.02726984024 s\n",
      "get_feature_locations  2.85375499725 s\n",
      "get_feature_locations  2.74376106262 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_3889\n",
      "gene_3991\n",
      "gene_3991\n",
      "get_mrna_by_gene  1.8807861805 s\n",
      "[u'mRNA_3890']\n",
      "get_cds_by_mrna  1.98944592476 s\n",
      "get_mrna_exons  1.74452996254 s\n",
      "get_feature_locations  2.15993118286 s\n",
      "get_feature_locations  2.43378210068 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_3890\n",
      "gene_3994\n",
      "gene_3994\n",
      "get_mrna_by_gene  2.11650681496 s\n",
      "[u'mRNA_3893']\n",
      "get_cds_by_mrna  1.99657201767 s\n",
      "get_mrna_exons  2.04512000084 s\n",
      "get_feature_locations  2.26067304611 s\n",
      "get_feature_locations  2.58654904366 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_3893\n",
      "gene_3995\n",
      "gene_3995\n",
      "get_mrna_by_gene  1.96572589874 s\n",
      "[u'mRNA_3894']\n",
      "get_cds_by_mrna  2.08969593048 s\n",
      "get_mrna_exons  1.37279701233 s\n",
      "get_feature_locations  2.15558600426 s\n",
      "get_feature_locations  2.35059690475 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_3894\n",
      "gene_3996\n",
      "gene_3996\n",
      "get_mrna_by_gene  2.07586693764 s\n",
      "[u'mRNA_3895']\n",
      "get_cds_by_mrna  1.783618927 s\n",
      "get_mrna_exons  1.84924888611 s\n",
      "get_feature_locations  2.37503790855 s\n",
      "get_feature_locations  2.13473510742 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_3895\n",
      "gene_3997\n",
      "gene_3997\n",
      "get_mrna_by_gene  1.37631201744 s\n",
      "[u'mRNA_3896']\n",
      "get_cds_by_mrna  1.45806407928 s\n",
      "get_mrna_exons  1.56818699837 s\n",
      "get_feature_locations  2.14966106415 s\n",
      "get_feature_locations  1.71960806847 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_3896\n",
      "gene_3998\n",
      "gene_3998\n",
      "get_mrna_by_gene  2.03419685364 s\n",
      "[u'mRNA_3897']\n",
      "get_cds_by_mrna  1.37596011162 s\n",
      "get_mrna_exons  1.35688209534 s\n",
      "get_feature_locations  3.05720305443 s\n",
      "get_feature_locations  2.76209497452 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_3897\n",
      "gene_3999\n",
      "gene_3999\n",
      "get_mrna_by_gene  1.70844888687 s\n",
      "[u'mRNA_3898']\n",
      "get_cds_by_mrna  1.72272515297 s\n",
      "get_mrna_exons  1.53422307968 s\n",
      "get_feature_locations  2.22487092018 s\n",
      "get_feature_locations  2.33454799652 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_3898\n",
      "gene_3992\n",
      "gene_3992\n",
      "get_mrna_by_gene  2.15234899521 s\n",
      "[u'mRNA_3891']\n",
      "get_cds_by_mrna  1.45831084251 s\n",
      "get_mrna_exons  1.75284385681 s\n",
      "get_feature_locations  2.89223814011 s\n",
      "get_feature_locations  2.35570120811 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_3891\n",
      "gene_4076\n",
      "gene_4076\n",
      "get_mrna_by_gene  1.28170394897 s\n",
      "[u'mRNA_3972']\n",
      "get_cds_by_mrna  1.30925583839 s\n",
      "get_mrna_exons  1.65246391296 s\n"
     ]
    },
    {
     "ename": "ConnectionError",
     "evalue": "HTTPSConnectionPool(host='ci.kbase.us', port=443): Max retries exceeded with url: /services/ws/ (Caused by <class 'socket.error'>: [Errno 50] Network is down)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mConnectionError\u001b[0m                           Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-9-ab8dd783a4bc>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mgenomeid\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"280699_Ensembl\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mwsthis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"ReferenceEnsemblPlantGenomeAnnotations\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mrunBig\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mws\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mwsthis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgenome_id\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgenomeid\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-4-5c4ab57a8020>\u001b[0m in \u001b[0;36mrunBig\u001b[0;34m(ws_url, ws, genome_id)\u001b[0m\n\u001b[1;32m    191\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mf\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfeat_data\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    192\u001b[0m             \u001b[0;32mprint\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 193\u001b[0;31m             \u001b[0mgffdata\u001b[0m\u001b[0;34m+=\u001b[0m\u001b[0mget_gff\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mgenome_annotation\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mtaxonid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msourceinfo\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    194\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    195\u001b[0m         \u001b[0;31m#break after first contig for now\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-8-2c602074b4a5>\u001b[0m in \u001b[0;36mget_gff\u001b[0;34m(gene, genome_annotation, taxon_id, source_info)\u001b[0m\n\u001b[1;32m     37\u001b[0m         \u001b[0;31m#retrieve locations for mRNAs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m         \u001b[0mt0\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 39\u001b[0;31m         \u001b[0mmrna_location\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgenome_annotation\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_feature_locations\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlistmrna\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     40\u001b[0m         \u001b[0mt1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     41\u001b[0m         \u001b[0;32mprint\u001b[0m \u001b[0;34m\"get_feature_locations \"\u001b[0m \u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mt1\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mt0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"s\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/doekbase/data_api/annotation/genome_annotation/api.pyc\u001b[0m in \u001b[0;36mget_feature_locations\u001b[0;34m(self, feature_id_list)\u001b[0m\n\u001b[1;32m    488\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    489\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mget_feature_locations\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfeature_id_list\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 490\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mproxy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_feature_locations\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfeature_id_list\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    491\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    492\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mget_feature_dna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfeature_id_list\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/doekbase/data_api/annotation/genome_annotation/api.pyc\u001b[0m in \u001b[0;36mget_feature_locations\u001b[0;34m(self, feature_id_list)\u001b[0m\n\u001b[1;32m   1325\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1326\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mget_feature_locations\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfeature_id_list\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1327\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_feature_data\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"locations\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfeature_id_list\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1328\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1329\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mget_feature_dna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfeature_id_list\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/doekbase/data_api/annotation/genome_annotation/api.pyc\u001b[0m in \u001b[0;36m_get_feature_data\u001b[0;34m(self, data, feature_id_list)\u001b[0m\n\u001b[1;32m   1288\u001b[0m                 \u001b[0mworking_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfeatures\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1289\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1290\u001b[0;31m                 \u001b[0mcontainer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mObjectAPI\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mservices\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_token\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mref\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1291\u001b[0m                 \u001b[0mfeatures\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcontainer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_data_subset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath_list\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfeature_refs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"features\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1292\u001b[0m                 \u001b[0mworking_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfeature_id_list\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/doekbase/data_api/core.pyc\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, services, token, ref)\u001b[0m\n\u001b[1;32m    128\u001b[0m             \u001b[0;34m\"objects\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m\"ref\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mref\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    129\u001b[0m             \u001b[0;34m\"includeMetadata\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 130\u001b[0;31m             \"ignoreErrors\": 0})\n\u001b[0m\u001b[1;32m    131\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0minfo_values\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    132\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Cannot find object: {}\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mref\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/doekbase/workspace/client.pyc\u001b[0m in \u001b[0;36mget_object_info_new\u001b[0;34m(self, params)\u001b[0m\n\u001b[1;32m    304\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mget_object_info_new\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparams\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    305\u001b[0m         resp = self._call('Workspace.get_object_info_new',\n\u001b[0;32m--> 306\u001b[0;31m                           [params])\n\u001b[0m\u001b[1;32m    307\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresp\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    308\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/doekbase/workspace/client.pyc\u001b[0m in \u001b[0;36m_call\u001b[0;34m(self, method, params)\u001b[0m\n\u001b[1;32m    149\u001b[0m         ret = _requests.post(self.url, data=body, headers=self._headers,\n\u001b[1;32m    150\u001b[0m                              \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 151\u001b[0;31m                              verify=not self.trust_all_ssl_certificates)\n\u001b[0m\u001b[1;32m    152\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mret\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstatus_code\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0m_requests\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcodes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mserver_error\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    153\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0m_CT\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mret\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mheaders\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mret\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mheaders\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0m_CT\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0m_AJ\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/requests/api.pyc\u001b[0m in \u001b[0;36mpost\u001b[0;34m(url, data, **kwargs)\u001b[0m\n\u001b[1;32m     86\u001b[0m     \"\"\"\n\u001b[1;32m     87\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 88\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mrequest\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'post'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     89\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     90\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/requests/api.pyc\u001b[0m in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     42\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     43\u001b[0m     \u001b[0msession\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msessions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSession\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 44\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0msession\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     45\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     46\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/requests/sessions.pyc\u001b[0m in \u001b[0;36mrequest\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert)\u001b[0m\n\u001b[1;32m    454\u001b[0m             \u001b[0;34m'allow_redirects'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mallow_redirects\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    455\u001b[0m         }\n\u001b[0;32m--> 456\u001b[0;31m         \u001b[0mresp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprep\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0msend_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    457\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    458\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/requests/sessions.pyc\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    557\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    558\u001b[0m         \u001b[0;31m# Send the request\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 559\u001b[0;31m         \u001b[0mr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0madapter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    560\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    561\u001b[0m         \u001b[0;31m# Total elapsed time of the request (approximately)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/requests/adapters.pyc\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    373\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    374\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mMaxRetryError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 375\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mConnectionError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrequest\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    376\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    377\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0m_ProxyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mConnectionError\u001b[0m: HTTPSConnectionPool(host='ci.kbase.us', port=443): Max retries exceeded with url: /services/ws/ (Caused by <class 'socket.error'>: [Errno 50] Network is down)"
     ]
    }
   ],
   "source": [
    "genomeid=\"280699_Ensembl\"\n",
    "wsthis = \"ReferenceEnsemblPlantGenomeAnnotations\"\n",
    "runBig(ws=wsthis, genome_id=genomeid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#write overall header\n",
    "#for contigs\n",
    "    #get contig features\n",
    "    #write contig header\n",
    "    #for gene in contig\n",
    "        #write GFF gene\n",
    "            #for mRNA from gene\n",
    "                #write GFF mRNA for gene\n",
    "                    #for exon in mRNA\n",
    "                        #write GFF exon for mRNA\n",
    "                        #write GFF CDS for mRNA\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
