{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import doekbase.data_api\n",
    "from doekbase.data_api.annotation.genome_annotation.api import GenomeAnnotationAPI\n",
    "import time\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_gff(gene,genome_annotation, taxon_id, source_info):\n",
    "    \n",
    "    print(gene)\n",
    "    #retrieve pairwise inter-feature relationships between genes, mRNAs, CDSs\n",
    "    #the below section of code has method timings torecord performance  \n",
    "    t0 = time.time()\n",
    "    listmrna=genome_annotation.get_mrna_by_gene([gene])[gene]    \n",
    "    t1 = time.time()\n",
    "    print \"get_mrna_by_gene \" , (t1-t0), \"s\"\n",
    "    #print(listmrna)\n",
    "    \n",
    "    #t0 = time.time()\n",
    "    #listcds=genome_annotation.get_cds_by_gene([gene])[gene] \n",
    "    #t1 = time.time()\n",
    "    #print \"get_cds_by_gene \" , (t1-t0), \"s\"   \n",
    "    #print(listcds)\n",
    "        \n",
    "    t0 = time.time()\n",
    "    print(listmrna)\n",
    "    mrna_cds=genome_annotation.get_cds_by_mrna(listmrna)\n",
    "    t1 = time.time()\n",
    "    print \"get_cds_by_mrna \" , (t1-t0), \"s\"\n",
    "    #print(mrna_cds)\n",
    "    \n",
    "    #t0 = time.time()\n",
    "    #cds_mrna=genome_annotation.get_mrna_by_cds(listcds)\n",
    "    #t1 = time.time()\n",
    "    #print \"get_mrna_by_cds \" , (t1-t0), \"s\"\n",
    "    \n",
    "    t0 = time.time()\n",
    "    get_mrna_exons = genome_annotation.get_mrna_exons(listmrna)\n",
    "    t1 = time.time()\n",
    "    print \"get_mrna_exons \" , (t1-t0), \"s\"\n",
    "    #print get_mrna_exons\n",
    "    \n",
    "    #retrieve locations for genes, mRNAs, and cds\n",
    "    t0 = time.time()\n",
    "    gene_location=genome_annotation.get_feature_locations([gene]) \n",
    "    t1 = time.time()\n",
    "    print \"get_feature_locations \" , (t1-t0), \"s\"\n",
    "\n",
    "    t0 = time.time()\n",
    "    mrna_location=genome_annotation.get_feature_locations(listmrna) \n",
    "    t1 = time.time()\n",
    "    print \"get_feature_locations \" , (t1-t0), \"s\"\n",
    "\n",
    "    #t0 = time.time()\n",
    "    #cds_location=genome_annotation.get_feature_locations(listcds) \n",
    "    #t1 = time.time()\n",
    "    #print \"get_feature_locations \" , (t1-t0), \"s\"\n",
    "    \n",
    "    string=\"\"\n",
    "    \n",
    "    string+=\"##sequence-region \"+gene_location[gene_location.keys()[0]][0]['contig_id']+\"\\t\"+\"start\"+\"\\t\"+\"stop\"+\"\\n\"##NC_003070.9 1 30427671\n",
    "    if taxon_id != -1:\n",
    "        string+=\"##species http://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=\"+str(taxon_id)+\"\\n\"#http://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=3702\n",
    "    else :\n",
    "        string+=\"##species unknown\\n\"\n",
    "    \n",
    "\n",
    "    contig_id = \"\"\n",
    "    #output the gene data\n",
    "    for x in gene_location:\n",
    "        #print(type(x))\n",
    "        #print(x)\n",
    "        info=gene_location[x][0]\n",
    "        #print(type(info))\n",
    "        #print info\n",
    "        stop=info['start'] + info['length']\n",
    "        \n",
    "        contig_id = info['contig_id']\n",
    "        \n",
    "        #print \"source_info\"\n",
    "        #print(type(source_info))\n",
    "        #print(source_info)\n",
    "        #output aliases for gene as dbxref?\n",
    "        string+=info['contig_id']  + \"\\t\" + source_info['external_source']+\"\\tgene\\t\"  + str(info['start']) + \"\\t\" + str(stop) + \"\\t.\\t\" +  info['strand']+ \"\\t.\\t\" + \"ID=\" + x + \";\" + \"Name=\"+ x +\"\\n\"\n",
    "    \n",
    "    print \"len(listmrna)\"    \n",
    "    print(len(listmrna))\n",
    "    if len(listmrna) > 0:    \n",
    "    #output the mRNA data\n",
    "        for mrna in listmrna:\n",
    "            print(\"mrna \"+mrna)\n",
    "            cds=mrna_cds[mrna]\n",
    "            infomrna=mrna_location[mrna][0]\n",
    "            #print(type(infomrna))\n",
    "            #print(infomrna)\n",
    "            #infocds=cds_location[mrna_cds[mrna]]\n",
    "            stop=infomrna['start'] + infomrna['length']            \n",
    "            ###output mrna data\n",
    "            string+=infomrna['contig_id']  + \"\\t\" +source_info['external_source']+\"\\tmRNA\\t\"  + str(infomrna['start'])+ \"\\t\" + str(stop) + \"\\t.\\t\" +  infomrna['strand']+ \"\\t.\\t\" + \"ID=\" + mrna + \";\" + \"Parent=\"+ gene+\"\\n\"\n",
    "            \n",
    "            #print 'get_mrna_exons.keys()'\n",
    "            #print get_mrna_exons.keys()\n",
    "            for mrna in get_mrna_exons.keys():\n",
    "                #print(type(get_mrna_exons[mrna]))\n",
    "                #print(get_mrna_exons[mrna])\n",
    "                #print \"get_mrna_exons[mrna][i]['exon_location']\"\n",
    "                #print type(get_mrna_exons[mrna][0]['exon_location'])\n",
    "                ###output exon data\n",
    "                for i in range(0, len(get_mrna_exons[mrna])):\n",
    "                    #print \"mrna, gene\"\n",
    "                    #print type(mrna)\n",
    "                    #print type(gene)\n",
    "                    stop= get_mrna_exons[mrna][i]['exon_location']['start'] + get_mrna_exons[mrna][i]['exon_location']['length']\n",
    "                    string+= get_mrna_exons[mrna][i]['exon_location']['contig_id']  + \"\\t\" +source_info['external_source'] + \"\\texon\\t\"  + str(get_mrna_exons[mrna][i]['exon_location']['start'])+ \"\\t\" + str(stop) + \"\\t.\\t\" +  get_mrna_exons[mrna][i]['exon_location']['strand']+ \"\\t.\\t\" + \"ID=\" + mrna +\".exon.\" + str(i)  + \";\" + \"Parent=\"+ mrna+\"\\n\"\n",
    "                ###output exon data\n",
    "                for i in range(0, len(get_mrna_exons[mrna])):\n",
    "                    stop= get_mrna_exons[mrna][i]['exon_location']['start'] + get_mrna_exons[mrna][i]['exon_location']['length']\n",
    "                    #***TODO: set PHASE for CDS\n",
    "                    string+= get_mrna_exons[mrna][i]['exon_location']['contig_id']  + \"\\t\" +source_info['external_source'] + \"\\tCDS\\t\"  + str(get_mrna_exons[mrna][i]['exon_location']['start'])+ \"\\t\" + str(stop) + \"\\t.\\t\" +  get_mrna_exons[mrna][i]['exon_location']['strand']+ \"\\t.\\t\" + \"ID=\" + mrna +\".CDS.\" + str(i)  + \";\" + \"Parent=\"+ mrna+\"\\n\"\n",
    "\n",
    "    return string       \n",
    "\n",
    "#hardcoded test with 5 genes\n",
    "def run(ws_url='https://ci.kbase.us/services/ws/', ws=\"ReferenceEnsemblPlantGenomeAnnotations\", genome_id=\"280699_Ensembl\"):\n",
    "    #an example KBase reference genome\n",
    "    genome_ref = \"ReferenceEnsemblPlantGenomeAnnotations\"+\"/\"+genome_id\n",
    "    #instantiate a new genome annotation API\n",
    "    genome_annotation = GenomeAnnotationAPI(services = {\"workspace_service_url\": ws_url}, token=os.environ.get('KB_AUTH_TOKEN'), ref=genome_ref)\n",
    "    taxon = genome_annotation.get_taxon()\n",
    "    taxonid = taxon.get_taxonomic_id()\n",
    "    \n",
    "    if taxonid == -1 :\n",
    "        taxonid = \"unknown\"\n",
    "    \n",
    "    assembly = genome_annotation.get_assembly()\n",
    "    contig_ids = assembly.get_contig_ids()\n",
    "    sourceinfo = assembly.get_external_source_info()\n",
    "    genes=['gene_1','gene_2','gene_5', 'gene_100', 'gene_1000']\n",
    "    #genes=['gene_1']\n",
    "    \n",
    "    gffdata=\"\"\n",
    "    gffdata+=\"##gff-version 3\\n\"\n",
    "    gffdata+=\"#!gff-spec-version 1.21\\n\"\n",
    "    gffdata+=\"#!processor KBase GFF3 downloader\\n\"\n",
    "    #!genome-build TAIR10\n",
    "    #!genome-build-accession NCBI_Assembly:GCF_000001735.3   \n",
    "    \n",
    "    for s in genes:\n",
    "        gffdata+=get_gff(s,genome_annotation,taxonid, sourceinfo)\n",
    "\n",
    "    outfile = genome_id+'.gff'\n",
    "    print outfile\n",
    "    with open(outfile, 'w') as f:\n",
    "        f.write(gffdata)\n",
    "    \n",
    "#\n",
    "def runBig(ws_url='https://ci.kbase.us/services/ws/', ws=\"ReferenceEnsemblPlantGenomeAnnotations\", genome_id=\"280699_Ensembl\"):\n",
    "    #an example KBase reference genome\n",
    "    genome_ref = \"ReferenceEnsemblPlantGenomeAnnotations\"+\"/\"+genome_id#kb|g.166819\"\n",
    "    #instantiate a new genome annotation API\n",
    "    genome_annotation = GenomeAnnotationAPI(services = {\"workspace_service_url\": ws_url}, token=os.environ.get('KB_AUTH_TOKEN'), ref=genome_ref)\n",
    "    taxon = genome_annotation.get_taxon()\n",
    "    taxonid = taxon.get_taxonomic_id()\n",
    "    \n",
    "    if taxonid == -1 :\n",
    "        taxonid = \"unknown\"\n",
    "    \n",
    "    assembly = genome_annotation.get_assembly()\n",
    "    contig_ids = assembly.get_contig_ids()\n",
    "    sourceinfo = assembly.get_external_source_info()\n",
    "    \n",
    "    gffdata=\"\"\n",
    "    gffdata+=\"##gff-version 3\\n\"\n",
    "    gffdata+=\"#!gff-spec-version 1.21\\n\"\n",
    "    gffdata+=\"#!processor KBase GFF3 downloader\\n\"\n",
    "    #!genome-build TAIR10\n",
    "    #!genome-build-accession NCBI_Assembly:GCF_000001735.3  \n",
    "    \n",
    "    start=1\n",
    "    stop=100000000\n",
    "    \n",
    "    feat_data = []\n",
    "    for c in contig_ids:\n",
    "        regions = []\n",
    "        contig_add = {\"contig_id\": c, \"start\": start, \"length\": stop-start, \"strand\": \"+\"}\n",
    "        contig_add2 = {\"contig_id\": c, \"start\": stop, \"length\": stop-start, \"strand\": \"-\"}\n",
    "        regions.append(contig_add)\n",
    "        regions.append(contig_add2)\n",
    "        feat_data = genome_annotation.get_feature_ids(filters={\"region_list\":regions,\"type_list\":[\"gene\"]})['by_type']['gene']\n",
    "        \n",
    "        print \"feat_data len \"+ str(len(feat_data))\n",
    "        #print feat_data\n",
    "        #for s in genes:\n",
    "        for f in feat_data:\n",
    "            print f\n",
    "            gffdata+=get_gff(f,genome_annotation,taxonid, sourceinfo)\n",
    "\n",
    "    string+=\"###\"\n",
    "    outfile = genome_id+'.gff'\n",
    "    print outfile\n",
    "    with open(outfile, 'w') as f:\n",
    "        f.write(gffdata)\n",
    "        \n",
    "    feat_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_1\n",
      "get_mrna_by_gene  1.67064189911 s\n",
      "[u'mRNA_1']\n",
      "get_cds_by_mrna  1.50494503975 s\n",
      "get_mrna_exons  1.49544405937 s\n",
      "get_feature_locations  1.95886707306 s\n",
      "get_feature_locations  2.66089892387 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_1\n",
      "gene_2\n",
      "get_mrna_by_gene  1.5816628933 s\n",
      "[u'mRNA_2']\n",
      "get_cds_by_mrna  1.58651709557 s\n",
      "get_mrna_exons  1.58610486984 s\n",
      "get_feature_locations  1.94034600258 s\n",
      "get_feature_locations  1.88635587692 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_2\n",
      "gene_5\n",
      "get_mrna_by_gene  1.54612207413 s\n",
      "[u'mRNA_3']\n",
      "get_cds_by_mrna  1.53069210052 s\n",
      "get_mrna_exons  1.67014789581 s\n",
      "get_feature_locations  2.21215605736 s\n",
      "get_feature_locations  2.13090801239 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_3\n",
      "gene_100\n",
      "get_mrna_by_gene  1.62239289284 s\n",
      "[u'mRNA_63']\n",
      "get_cds_by_mrna  1.63559103012 s\n",
      "get_mrna_exons  1.58180403709 s\n",
      "get_feature_locations  1.98508691788 s\n",
      "get_feature_locations  2.41667795181 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_63\n",
      "gene_1000\n",
      "get_mrna_by_gene  1.99698400497 s\n",
      "[u'mRNA_932']\n",
      "get_cds_by_mrna  1.54204797745 s\n",
      "get_mrna_exons  1.55971193314 s\n",
      "get_feature_locations  2.11794495583 s\n",
      "get_feature_locations  2.13708090782 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_932\n",
      "280699_Ensembl.gff\n"
     ]
    }
   ],
   "source": [
    "genomeid=\"280699_Ensembl\"\n",
    "wsthis = \"ReferenceEnsemblPlantGenomeAnnotations\"\n",
    "run(ws=wsthis, genome_id=genomeid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "feat_data len 363\n",
      "gene_3990\n",
      "gene_3990\n",
      "get_mrna_by_gene  1.55253696442 s\n",
      "[u'mRNA_3889']\n",
      "get_cds_by_mrna  1.54010915756 s\n",
      "get_mrna_exons  2.20199990273 s\n",
      "get_feature_locations  3.07964801788 s\n",
      "get_feature_locations  2.78345012665 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_3889\n",
      "gene_3991\n",
      "gene_3991\n",
      "get_mrna_by_gene  1.90872597694 s\n",
      "[u'mRNA_3890']\n",
      "get_cds_by_mrna  2.27354383469 s\n",
      "get_mrna_exons  2.11737895012 s\n",
      "get_feature_locations "
     ]
    }
   ],
   "source": [
    "genomeid=\"280699_Ensembl\"\n",
    "wsthis = \"ReferenceEnsemblPlantGenomeAnnotations\"\n",
    "runBig(ws=wsthis, genome_id=genomeid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#write overall header\n",
    "#for contigs\n",
    "    #get contig features\n",
    "    #write contig header\n",
    "    #for gene in contig\n",
    "        #write GFF gene\n",
    "            #for mRNA from gene\n",
    "                #write GFF mRNA for gene\n",
    "                    #for exons in mRNA\n",
    "                        #write GFF exon for mRNA\n",
    "                        #write GFF CDS for mRNA\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
