{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import doekbase.data_api\n",
    "from doekbase.data_api.annotation.genome_annotation.api import GenomeAnnotationAPI\n",
    "import time\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def get_gff(gene,genome_annotation, taxon_id, source_info, count, contig_len):\n",
    "    \n",
    "    print(gene)\n",
    "    #retrieve pairwise inter-feature relationships between genes, mRNAs, CDSs\n",
    "    #the below section of code has method timings torecord performance  \n",
    "    t0 = time.time()\n",
    "    listmrna=genome_annotation.get_mrna_by_gene([gene])[gene]    \n",
    "    t1 = time.time()\n",
    "    print \"get_mrna_by_gene \" , (t1-t0), \"s\"\n",
    "    #print(listmrna)\n",
    "    \n",
    "    #t0 = time.time()\n",
    "    #listcds=genome_annotation.get_cds_by_gene([gene])[gene] \n",
    "    #t1 = time.time()\n",
    "    #print \"get_cds_by_gene \" , (t1-t0), \"s\"   \n",
    "    #print(listcds)\n",
    "        \n",
    "    if len(listmrna) > 0:\n",
    "        t0 = time.time()\n",
    "        print(listmrna)\n",
    "        mrna_cds=genome_annotation.get_cds_by_mrna(listmrna)\n",
    "        t1 = time.time()\n",
    "        print \"get_cds_by_mrna \" , (t1-t0), \"s\"\n",
    "        #print(mrna_cds)\n",
    "    \n",
    "        #t0 = time.time()\n",
    "        #cds_mrna=genome_annotation.get_mrna_by_cds(listcds)\n",
    "        #t1 = time.time()\n",
    "        #print \"get_mrna_by_cds \" , (t1-t0), \"s\"\n",
    "\n",
    "        t0 = time.time()\n",
    "        get_mrna_exons = genome_annotation.get_mrna_exons(listmrna)\n",
    "        t1 = time.time()\n",
    "        print \"get_mrna_exons \" , (t1-t0), \"s\"\n",
    "        #print get_mrna_exons\n",
    "        \n",
    "        #retrieve locations for mRNAs\n",
    "        t0 = time.time()\n",
    "        mrna_location=genome_annotation.get_feature_locations(listmrna) \n",
    "        t1 = time.time()\n",
    "        print \"get_feature_locations \" , (t1-t0), \"s\"\n",
    "\n",
    "\n",
    "    #retrieve locations for genes, mRNAs, and cds\n",
    "    t0 = time.time()\n",
    "    gene_location=genome_annotation.get_feature_locations([gene]) \n",
    "    t1 = time.time()\n",
    "    print \"get_feature_locations \" , (t1-t0), \"s\"\n",
    "    \n",
    "    #t0 = time.time()\n",
    "    #cds_location=genome_annotation.get_feature_locations(listcds) \n",
    "    #t1 = time.time()\n",
    "    #print \"get_feature_locations \" , (t1-t0), \"s\"\n",
    "    \n",
    "    string=\"\"\n",
    "    \n",
    "    if count == 0:\n",
    "        string+=\"##sequence-region \"+gene_location[gene_location.keys()[0]][0]['contig_id']+\"\\t\"+\"1\"+\"\\t\"+str(contig_len)+\"\\n\"##NC_003070.9 1 30427671\n",
    "        if taxon_id != -1:\n",
    "            string+=\"##species http://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=\"+str(taxon_id)+\"\\n\"#http://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=3702\n",
    "        else :\n",
    "            string+=\"##species unknown\\n\"\n",
    "    \n",
    "\n",
    "    contig_id = \"\"\n",
    "    #output the gene data\n",
    "    for x in gene_location:\n",
    "        #print(type(x))\n",
    "        #print(x)\n",
    "        info=gene_location[x][0]\n",
    "        #print(type(info))\n",
    "        #print info\n",
    "        stop=info['start'] + info['length']\n",
    "        \n",
    "        contig_id = info['contig_id']\n",
    "        \n",
    "        #print \"source_info\"\n",
    "        #print(type(source_info))\n",
    "        #print(source_info)\n",
    "        #output aliases for gene as dbxref?\n",
    "        string+=info['contig_id']  + \"\\t\" + source_info['external_source']+\"\\tgene\\t\"  + str(info['start']) + \"\\t\" + str(stop) + \"\\t.\\t\" +  info['strand']+ \"\\t.\\t\" + \"ID=\" + x + \";\" + \"Name=\"+ x +\"\\n\"\n",
    "    \n",
    "    print \"len(listmrna)\"    \n",
    "    print(len(listmrna))\n",
    "    if len(listmrna) > 0:    \n",
    "    #output the mRNA data\n",
    "        for mrna in listmrna:\n",
    "            print(\"mrna \"+mrna)\n",
    "            cds=mrna_cds[mrna]\n",
    "            infomrna=mrna_location[mrna][0]\n",
    "            #print(type(infomrna))\n",
    "            #print(infomrna)\n",
    "            #infocds=cds_location[mrna_cds[mrna]]\n",
    "            stop=infomrna['start'] + infomrna['length']            \n",
    "            ###output mrna data\n",
    "            string+=infomrna['contig_id']  + \"\\t\" +source_info['external_source']+\"\\tmRNA\\t\"  + str(infomrna['start'])+ \"\\t\" + str(stop) + \"\\t.\\t\" +  infomrna['strand']+ \"\\t.\\t\" + \"ID=\" + mrna + \";\" + \"Parent=\"+ gene+\"\\n\"\n",
    "            \n",
    "            #print 'get_mrna_exons.keys()'\n",
    "            #print get_mrna_exons.keys()\n",
    "            \n",
    "            ##TODO output UTRs\n",
    "            for mrna in get_mrna_exons.keys():\n",
    "                #print(type(get_mrna_exons[mrna]))\n",
    "                #print(get_mrna_exons[mrna])\n",
    "                #print \"get_mrna_exons[mrna][i]['exon_location']\"\n",
    "                #print type(get_mrna_exons[mrna][0]['exon_location'])\n",
    "                ###output exon data\n",
    "                for i in range(0, len(get_mrna_exons[mrna])):\n",
    "                    #print \"mrna, gene\"\n",
    "                    #print type(mrna)\n",
    "                    #print type(gene)\n",
    "                    stop= get_mrna_exons[mrna][i]['exon_location']['start'] + get_mrna_exons[mrna][i]['exon_location']['length']\n",
    "                    string+= get_mrna_exons[mrna][i]['exon_location']['contig_id']  + \"\\t\" +source_info['external_source'] + \"\\texon\\t\"  + str(get_mrna_exons[mrna][i]['exon_location']['start'])+ \"\\t\" + str(stop) + \"\\t.\\t\" +  get_mrna_exons[mrna][i]['exon_location']['strand']+ \"\\t.\\t\" + \"ID=\" + mrna +\".exon.\" + str(i)  + \";\" + \"Parent=\"+ mrna+\"\\n\"\n",
    "                ###output CDS data\n",
    "                for i in range(0, len(get_mrna_exons[mrna])):\n",
    "                    stop= get_mrna_exons[mrna][i]['exon_location']['start'] + get_mrna_exons[mrna][i]['exon_location']['length']\n",
    "                    ##***TODO: set PHASE for CDS\n",
    "                    string+= get_mrna_exons[mrna][i]['exon_location']['contig_id']  + \"\\t\" +source_info['external_source'] + \"\\tCDS\\t\"  + str(get_mrna_exons[mrna][i]['exon_location']['start'])+ \"\\t\" + str(stop) + \"\\t.\\t\" +  get_mrna_exons[mrna][i]['exon_location']['strand']+ \"\\t.\\t\" + \"ID=\" + mrna +\".CDS.\" + str(i)  + \";\" + \"Parent=\"+ mrna+\"\\n\"\n",
    "\n",
    "    return string       \n",
    "\n",
    "#hardcoded test with 5 random genes\n",
    "def runDemo(ws_url='https://ci.kbase.us/services/ws/', ws=\"ReferenceEnsemblPlantGenomeAnnotations\", genome_id=\"280699_Ensembl\"):\n",
    "    print \"1\"\n",
    "    #an example KBase reference genome\n",
    "    genome_ref = \"ReferenceEnsemblPlantGenomeAnnotations\"+\"/\"+genome_id\n",
    "    #instantiate a new genome annotation API\n",
    "    t0 = time.time()\n",
    "    genome_annotation = GenomeAnnotationAPI(services = {\"workspace_service_url\": ws_url}, token=os.environ.get('KB_AUTH_TOKEN'), ref=genome_ref)\n",
    "    t1 = time.time()\n",
    "    print \"GenomeAnnotationAPI \" , (t1-t0), \"s\"\n",
    "    \n",
    "    t0 = time.time()\n",
    "    taxon = genome_annotation.get_taxon()\n",
    "    t1 = time.time()\n",
    "    print \"get_taxon \" , (t1-t0), \"s\"\n",
    "        \n",
    "    t0 = time.time()\n",
    "    taxonid = taxon.get_taxonomic_id()\n",
    "    t1 = time.time()\n",
    "    print \"get_taxonomic_id \" , (t1-t0), \"s\"\n",
    "    \n",
    "    if taxonid == -1 :\n",
    "        taxonid = \"unknown\"\n",
    "    \n",
    "    assembly = genome_annotation.get_assembly()\n",
    "    contig_ids = assembly.get_contig_ids()\n",
    "    sourceinfo = assembly.get_external_source_info()\n",
    "    genes=['gene_1','gene_2','gene_5', 'gene_100', 'gene_1000']\n",
    "    #genes=['gene_1']\n",
    "    \n",
    "    gffdata=\"\"\n",
    "    gffdata+=\"##gff-version 3\\n\"\n",
    "    gffdata+=\"#!gff-spec-version 1.21\\n\"\n",
    "    gffdata+=\"#!processor KBase GFF3 downloader\\n\"\n",
    "    #!genome-build TAIR10\n",
    "    #!genome-build-accession NCBI_Assembly:GCF_000001735.3   \n",
    "    \n",
    "    count = 0\n",
    "    for s in genes:\n",
    "        gffdata+=get_gff(s,genome_annotation,taxonid, sourceinfo, count, -1)\n",
    "        count = count + 1\n",
    "\n",
    "    gffdata+=\"###\"\n",
    "    outfile = genome_id+'.gff'\n",
    "    print outfile\n",
    "    with open(outfile, 'w') as f:\n",
    "        f.write(gffdata)\n",
    "    \n",
    "#create GFF for an entire GenomeAnnotation\n",
    "def runGenome(ws_url='https://ci.kbase.us/services/ws/', ws=\"ReferenceEnsemblPlantGenomeAnnotations\", genome_id=\"280699_Ensembl\"):\n",
    "    #an example KBase reference genome\n",
    "    genome_ref = \"ReferenceEnsemblPlantGenomeAnnotations\"+\"/\"+genome_id#kb|g.166819\"\n",
    "    #instantiate a new genome annotation API\n",
    "    t0 = time.time()\n",
    "    genome_annotation = GenomeAnnotationAPI(services = {\"workspace_service_url\": ws_url}, token=os.environ.get('KB_AUTH_TOKEN'), ref=genome_ref)\n",
    "    t1 = time.time()\n",
    "    print \"GenomeAnnotationAPI \" , (t1-t0), \"s\"\n",
    "    \n",
    "    t0 = time.time()\n",
    "    taxon = genome_annotation.get_taxon()\n",
    "    t1 = time.time()\n",
    "    print \"get_taxon \" , (t1-t0), \"s\"\n",
    "        \n",
    "    t0 = time.time()\n",
    "    taxonid = taxon.get_taxonomic_id()\n",
    "    t1 = time.time()\n",
    "    print \"get_taxonomic_id \" , (t1-t0), \"s\"\n",
    "    \n",
    "    print \"1\"\n",
    "    \n",
    "    if taxonid == -1 :\n",
    "        taxonid = \"unknown\"\n",
    "    \n",
    "    assembly = genome_annotation.get_assembly()\n",
    "    contig_ids = assembly.get_contig_ids()\n",
    "    sourceinfo = assembly.get_external_source_info()\n",
    "    \n",
    "    contig_lens = assembly.get_contig_lengths()\n",
    "    #print contig_lens\n",
    "    \n",
    "    print \"2\"\n",
    "    \n",
    "    gffdata=\"\"\n",
    "    gffdata+=\"##gff-version 3\\n\"\n",
    "    gffdata+=\"#!gff-spec-version 1.21\\n\"\n",
    "    gffdata+=\"#!processor KBase GFF3 downloader\\n\"\n",
    "    #!genome-build TAIR10\n",
    "    #!genome-build-accession NCBI_Assembly:GCF_000001735.3  \n",
    "    \n",
    "    start=1\n",
    "    stop=100000000\n",
    "    \n",
    "    feat_data = []\n",
    "    print \"contig_ids len \"+ str(len(contig_ids))\n",
    "    countc = 0\n",
    "    feat_total = 0\n",
    "    for c in contig_ids:\n",
    "        \n",
    "        stop = contig_lens[c]\n",
    "        print c\n",
    "        regions = []\n",
    "        contig_add = {\"contig_id\": c, \"start\": start, \"length\": stop-start, \"strand\": \"+\"}\n",
    "        contig_add2 = {\"contig_id\": c, \"start\": stop, \"length\": stop-start, \"strand\": \"-\"}\n",
    "        regions.append(contig_add)\n",
    "        regions.append(contig_add2)\n",
    "        \n",
    "        t0 = time.time()\n",
    "        feat_data = genome_annotation.get_feature_ids(filters={\"region_list\":regions,\"type_list\":[\"gene\"]})['by_type']['gene']\n",
    "        t1 = time.time()\n",
    "        print \"get_feature_ids \" , (t1-t0), \"s\"\n",
    "    \n",
    "        print \"feat_data len \"+ str(len(feat_data))\n",
    "        #print feat_data\n",
    "        #for s in genes:\n",
    "        countf =0\n",
    "        feat_total = feat_total + len(feat_data)\n",
    "        for f in feat_data:\n",
    "            print f\n",
    "            gffdata+=get_gff(f,genome_annotation,taxonid, sourceinfo, countf, stop)\n",
    "            #print gffdata\n",
    "            countf = countf + 1\n",
    "            #break after small number       \n",
    "            if countf > 2:\n",
    "                print \"break after first 3 genes for now\"\n",
    "                break\n",
    "               \n",
    "        countc = countc + 1\n",
    "        if countc > 2:\n",
    "            print \"break after first 3 contigs for now \"   \n",
    "            #break after first contig for now    \n",
    "            break\n",
    "    \n",
    "    #print \"END\"\n",
    "    #print gffdata\n",
    "    \n",
    "    print \"process contigs \" + str(countc)\n",
    "    print \"feat total \" + str(feat_total)\n",
    "    \n",
    "    gffdata+=\"###\"\n",
    "    outfile = genome_id+'.gff'\n",
    "    print outfile\n",
    "    with open(outfile, 'w') as f:\n",
    "        f.write(gffdata)\n",
    "        \n",
    "    feat_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "GenomeAnnotationAPI  4.83470010757 s\n",
      "get_taxon  3.01157689095 s\n",
      "get_taxonomic_id  2.71797180176e-05 s\n",
      "gene_1\n",
      "get_mrna_by_gene  2.03141212463 s\n",
      "[u'mRNA_1']\n",
      "get_cds_by_mrna  2.27638292313 s\n",
      "get_mrna_exons  2.40531301498 s\n",
      "get_feature_locations  3.17668700218 s\n",
      "get_feature_locations  3.62060785294 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_1\n",
      "gene_2\n",
      "get_mrna_by_gene  2.80000805855 s\n",
      "[u'mRNA_2']\n",
      "get_cds_by_mrna  2.54419612885 s\n",
      "get_mrna_exons  3.1331140995 s\n",
      "get_feature_locations  3.34578895569 s\n",
      "get_feature_locations  3.85967493057 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_2\n",
      "gene_5\n",
      "get_mrna_by_gene  3.37133693695 s\n",
      "[u'mRNA_3']\n",
      "get_cds_by_mrna  2.73723483086 s\n",
      "get_mrna_exons  3.23955893517 s\n",
      "get_feature_locations  3.27452707291 s\n",
      "get_feature_locations  5.13207197189 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_3\n",
      "gene_100\n",
      "get_mrna_by_gene  4.27790999413 s\n",
      "[u'mRNA_63']\n",
      "get_cds_by_mrna  3.72632789612 s\n",
      "get_mrna_exons  3.1912920475 s\n",
      "get_feature_locations  3.42246103287 s\n",
      "get_feature_locations  2.94254207611 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_63\n",
      "gene_1000\n",
      "get_mrna_by_gene  2.06372594833 s\n",
      "[u'mRNA_932']\n",
      "get_cds_by_mrna  2.4581091404 s\n",
      "get_mrna_exons  3.16964888573 s\n",
      "get_feature_locations  3.10393309593 s\n",
      "get_feature_locations  5.1964700222 s\n",
      "len(listmrna)\n",
      "1\n",
      "mrna mRNA_932\n",
      "280699_Ensembl.gff\n"
     ]
    }
   ],
   "source": [
    "genomeid=\"280699_Ensembl\"\n",
    "wsthis = \"ReferenceEnsemblPlantGenomeAnnotations\"\n",
    "###Note that this demo function does not produce biologically valid GFF\n",
    "runDemo(ws=wsthis, genome_id=genomeid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ConnectionError",
     "evalue": "HTTPSConnectionPool(host='ci.kbase.us', port=443): Max retries exceeded with url: /services/ws/ (Caused by <class 'socket.gaierror'>: [Errno 8] nodename nor servname provided, or not known)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mConnectionError\u001b[0m                           Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-47-0aa5636ed70e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mgenomeid\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"280699_Ensembl\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mwsthis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"ReferenceEnsemblPlantGenomeAnnotations\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mrunGenome\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mws\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mwsthis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgenome_id\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgenomeid\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-46-fd69285270c9>\u001b[0m in \u001b[0;36mrunGenome\u001b[0;34m(ws_url, ws, genome_id)\u001b[0m\n\u001b[1;32m    174\u001b[0m     \u001b[0;31m#instantiate a new genome annotation API\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    175\u001b[0m     \u001b[0mt0\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 176\u001b[0;31m     \u001b[0mgenome_annotation\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mGenomeAnnotationAPI\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mservices\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m\"workspace_service_url\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mws_url\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtoken\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0menviron\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'KB_AUTH_TOKEN'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mref\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgenome_ref\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    177\u001b[0m     \u001b[0mt1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    178\u001b[0m     \u001b[0;32mprint\u001b[0m \u001b[0;34m\"GenomeAnnotationAPI \"\u001b[0m \u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mt1\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mt0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"s\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/doekbase/data_api/annotation/genome_annotation/api.pyc\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, services, token, ref)\u001b[0m\n\u001b[1;32m    454\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    455\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mservices\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtoken\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mref\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 456\u001b[0;31m         \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mGenomeAnnotationAPI\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mservices\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtoken\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mref\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    457\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    458\u001b[0m         \u001b[0mgeneric_object\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mObjectAPI\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mservices\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtoken\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mref\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/doekbase/data_api/core.pyc\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, services, token, ref)\u001b[0m\n\u001b[1;32m    128\u001b[0m             \u001b[0;34m\"objects\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m\"ref\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mref\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    129\u001b[0m             \u001b[0;34m\"includeMetadata\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 130\u001b[0;31m             \"ignoreErrors\": 0})\n\u001b[0m\u001b[1;32m    131\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0minfo_values\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    132\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Cannot find object: {}\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mref\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/doekbase/workspace/client.pyc\u001b[0m in \u001b[0;36mget_object_info_new\u001b[0;34m(self, params)\u001b[0m\n\u001b[1;32m    304\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mget_object_info_new\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparams\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    305\u001b[0m         resp = self._call('Workspace.get_object_info_new',\n\u001b[0;32m--> 306\u001b[0;31m                           [params])\n\u001b[0m\u001b[1;32m    307\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresp\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    308\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/doekbase/workspace/client.pyc\u001b[0m in \u001b[0;36m_call\u001b[0;34m(self, method, params)\u001b[0m\n\u001b[1;32m    149\u001b[0m         ret = _requests.post(self.url, data=body, headers=self._headers,\n\u001b[1;32m    150\u001b[0m                              \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 151\u001b[0;31m                              verify=not self.trust_all_ssl_certificates)\n\u001b[0m\u001b[1;32m    152\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mret\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstatus_code\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0m_requests\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcodes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mserver_error\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    153\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0m_CT\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mret\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mheaders\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mret\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mheaders\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0m_CT\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0m_AJ\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/requests/api.pyc\u001b[0m in \u001b[0;36mpost\u001b[0;34m(url, data, **kwargs)\u001b[0m\n\u001b[1;32m     86\u001b[0m     \"\"\"\n\u001b[1;32m     87\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 88\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mrequest\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'post'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     89\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     90\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/requests/api.pyc\u001b[0m in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     42\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     43\u001b[0m     \u001b[0msession\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msessions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSession\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 44\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0msession\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     45\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     46\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/requests/sessions.pyc\u001b[0m in \u001b[0;36mrequest\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert)\u001b[0m\n\u001b[1;32m    454\u001b[0m             \u001b[0;34m'allow_redirects'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mallow_redirects\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    455\u001b[0m         }\n\u001b[0;32m--> 456\u001b[0;31m         \u001b[0mresp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprep\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0msend_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    457\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    458\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/requests/sessions.pyc\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    557\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    558\u001b[0m         \u001b[0;31m# Send the request\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 559\u001b[0;31m         \u001b[0mr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0madapter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    560\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    561\u001b[0m         \u001b[0;31m# Total elapsed time of the request (approximately)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/requests/adapters.pyc\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    373\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    374\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mMaxRetryError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 375\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mConnectionError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrequest\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    376\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    377\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0m_ProxyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mConnectionError\u001b[0m: HTTPSConnectionPool(host='ci.kbase.us', port=443): Max retries exceeded with url: /services/ws/ (Caused by <class 'socket.gaierror'>: [Errno 8] nodename nor servname provided, or not known)"
     ]
    }
   ],
   "source": [
    "genomeid=\"280699_Ensembl\"\n",
    "wsthis = \"ReferenceEnsemblPlantGenomeAnnotations\"\n",
    "runGenome(ws=wsthis, genome_id=genomeid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#write overall header\n",
    "#for contigs\n",
    "    #get contig features\n",
    "    #write contig header\n",
    "    #for gene in contig\n",
    "        #write GFF gene\n",
    "            #for mRNA from gene\n",
    "                #write GFF mRNA for gene\n",
    "                    #for exon in mRNA\n",
    "                        #write GFF exon for mRNA\n",
    "                        #write GFF CDS for mRNA\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
