{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "\n",
       "    <script type=\"text/javascript\">\n",
       "      \n",
       "      (function(global) {\n",
       "        function now() {\n",
       "          return new Date();\n",
       "        }\n",
       "      \n",
       "        if (typeof (window._bokeh_onload_callbacks) === \"undefined\") {\n",
       "          window._bokeh_onload_callbacks = [];\n",
       "        }\n",
       "      \n",
       "        function run_callbacks() {\n",
       "          window._bokeh_onload_callbacks.forEach(function(callback) { callback() });\n",
       "          delete window._bokeh_onload_callbacks\n",
       "          console.info(\"Bokeh: all callbacks have finished\");\n",
       "        }\n",
       "      \n",
       "        function load_libs(js_urls, callback) {\n",
       "          window._bokeh_onload_callbacks.push(callback);\n",
       "          if (window._bokeh_is_loading > 0) {\n",
       "            console.log(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "            return null;\n",
       "          }\n",
       "          if (js_urls == null || js_urls.length === 0) {\n",
       "            run_callbacks();\n",
       "            return null;\n",
       "          }\n",
       "          console.log(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "          window._bokeh_is_loading = js_urls.length;\n",
       "          for (var i = 0; i < js_urls.length; i++) {\n",
       "            var url = js_urls[i];\n",
       "            var s = document.createElement('script');\n",
       "            s.src = url;\n",
       "            s.async = false;\n",
       "            s.onreadystatechange = s.onload = function() {\n",
       "              window._bokeh_is_loading--;\n",
       "              if (window._bokeh_is_loading === 0) {\n",
       "                console.log(\"Bokeh: all BokehJS libraries loaded\");\n",
       "                run_callbacks()\n",
       "              }\n",
       "            };\n",
       "            s.onerror = function() {\n",
       "              console.warn(\"failed to load library \" + url);\n",
       "            };\n",
       "            console.log(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "            document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          }\n",
       "        };var js_urls = ['https://cdn.pydata.org/bokeh/release/bokeh-0.11.0.min.js', 'https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.11.0.min.js', 'https://cdn.pydata.org/bokeh/release/bokeh-compiler-0.11.0.min.js'];\n",
       "      \n",
       "        var inline_js = [\n",
       "          function(Bokeh) {\n",
       "            Bokeh.set_log_level(\"info\");\n",
       "          },\n",
       "          function(Bokeh) {\n",
       "            console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-0.11.0.min.css\");\n",
       "            Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-0.11.0.min.css\");\n",
       "            console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.11.0.min.css\");\n",
       "            Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.11.0.min.css\");\n",
       "          }\n",
       "        ];\n",
       "      \n",
       "        function run_inline_js() {\n",
       "          for (var i = 0; i < inline_js.length; i++) {\n",
       "            inline_js[i](window.Bokeh);\n",
       "          }\n",
       "        }\n",
       "      \n",
       "        if (window._bokeh_is_loading === 0) {\n",
       "          console.log(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "          run_inline_js();\n",
       "        } else {\n",
       "          load_libs(js_urls, function() {\n",
       "            console.log(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "            run_inline_js();\n",
       "          });\n",
       "        }\n",
       "      }(this));\n",
       "    </script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import bokeh.charts\n",
    "import bokeh.plotting\n",
    "import bokeh.io\n",
    "from bokeh.plotting import figure, show, gridplot\n",
    "bokeh.io.output_notebook(hide_banner=True)\n",
    "\n",
    "from scipy.interpolate import spline\n",
    "\n",
    "import jinja2\n",
    "import IPython.display\n",
    "\n",
    "import ssl\n",
    "\n",
    "import numpy as np\n",
    "import scipy.special, scipy.stats\n",
    "\n",
    "import pprint\n",
    "import datetime\n",
    "import operator\n",
    "\n",
    "import os\n",
    "\n",
    "genomeannotationurl='https://ci.kbase.us/services/data/genome_annotation'\n",
    "assemblyurl='https://ci.kbase.us/services/data/assembly'\n",
    "\n",
    "token = os.environ[\"KB_AUTH_TOKEN\"]\n",
    "\n",
    "#from doekbase.data_api.annotation.genome_annotation import GenomeAnnotationAPI\n",
    "from doekbase.data_api.annotation.genome_annotation.api import GenomeAnnotationClientAPI\n",
    "from doekbase.data_api.sequence.assembly.api import AssemblyClientAPI\n",
    "\n",
    "services = {\"workspace_service_url\": \"https://ci.kbase.us/services/ws/\",\n",
    "                \"shock_service_url\": \"https://ci.kbase.us/services/shock_api/\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_genome_summary(genome_annotation=None):\n",
    "    \n",
    "    if genome_annotation == None:\n",
    "        raise TypeError(\"No GenomeAnnotation object given.\")\n",
    "    #elif genome_annotation.get_typestring().split('-')[0] not in biokbase.data_api.genome_annotation.TYPES:\n",
    "    #    raise TypeError(\"{0} is not a recognized GenomeAnnotation type.\".format(type(genome_annotation)))\n",
    "        \n",
    "    taxon = genome_annotation.get_taxon()\n",
    "    print \"genome_annotation.get_taxon \"+taxon\n",
    "    assemblyref = genome_annotation.get_assembly()\n",
    "    assemblyapi = AssemblyClientAPI(assemblyurl, token, assemblyref)\n",
    "    print \"genome_annotation.get_assembly\"\n",
    "    \n",
    "    overview = dict()\n",
    "    \n",
    "    #get assembly info\n",
    "    assembly_details = dict()\n",
    "    assembly_details[\"number_of_contigs\"] = assemblyapi.get_number_contigs()\n",
    "    print \"assembly.get_number_contigs\"\n",
    "    assembly_details[\"contig_length\"] = assemblyapi.get_contig_lengths()\n",
    "    print \"assembly.get_contig_lengths\"\n",
    "    assembly_details[\"contig_gc_content\"] = assemblyapi.get_contig_gc_content()\n",
    "    print \"assembly.get_contig_gc_content\"\n",
    "    \n",
    "    overview[\"assembly\"] = assembly_details\n",
    "    \n",
    "    #get annotation info\n",
    "    annotation_details = dict()\n",
    "    annotation_details[\"feature_types\"] = genome_annotation.get_feature_types()\n",
    "    \n",
    "    print annotation_details[\"feature_types\"]\n",
    "    \n",
    "    #remove 'exon' type which causes error for get_feature_type_descriptions()\n",
    "    del annotation_details[\"feature_types\"][-3]\n",
    "    \n",
    "    print annotation_details[\"feature_types\"]\n",
    "    \n",
    "    print \"genome_annotation.get_feature_types\"\n",
    "    annotation_details[\"feature_type_descriptions\"] = genome_annotation.get_feature_type_descriptions(annotation_details[\"feature_types\"])\n",
    "    print \"genome_annotation.get_feature_type_descriptions\"\n",
    "    annotation_details[\"feature_type_counts\"] = genome_annotation.get_feature_type_counts(annotation_details[\"feature_types\"])\n",
    "    print \"genome_annotation.get_feature_type_counts\"\n",
    "    \n",
    "    overview[\"annotation\"] = annotation_details\n",
    "    \n",
    "    return overview\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def plot_genome_quality(contig_gc_values=None,contig_len_values=None,feature_counts=None):\n",
    "    from bokeh.plotting import figure, show, gridplot\n",
    "    import numpy as np\n",
    "\n",
    "    log10_feature_counts = np.log10(feature_counts)    \n",
    "    log10_contig_len_values = np.log10(contig_len_values)\n",
    "    log10_contig_gc_values = np.log10(contig_gc_values)     \n",
    "\n",
    "        \n",
    "    global_scale = 130.0\n",
    "        \n",
    "    ### \n",
    "    xscale1 = len(contig_gc_values)\n",
    "    radius_scale1 = xscale1/global_scale\n",
    "    p1 = figure(title=\"Contig %GC by contig index\",background_fill=\"#FFF5E0\",plot_width=400, plot_height=400,title_text_font_size='14pt')\n",
    "\n",
    "    p1.scatter(np.arange(1,len(contig_gc_values)), contig_gc_values, marker=\"circle\",radius=log10_feature_counts*radius_scale1,\n",
    "                line_color=\"#6666ee\", fill_color=\"#ee6666\", fill_alpha=0.3, size=7, legend=\"Number of genes\")\n",
    "    p1.xaxis.axis_label=\"Contig index\"\n",
    "    p1.yaxis.axis_label=\"%GC\"\n",
    "        \n",
    "    global_scale1 = xscale1/radius_scale1\n",
    "    #115.5 (old)\n",
    "    \n",
    "    ###\n",
    "    from scipy.interpolate import spline\n",
    "\n",
    "    p2 = figure(title=\"Cumulative %GC distribution\",background_fill=\"#FFF5E0\",plot_width=400, plot_height=400,title_text_font_size='14pt')\n",
    "    x = np.cumsum(np.sort(contig_gc_values))/np.sum(contig_gc_values)\n",
    "    y = np.sort(contig_gc_values)\n",
    "\n",
    "    x_smooth = np.linspace(x.min(), x.max(), 50)\n",
    "    y_smooth = spline(x, y, x_smooth)\n",
    "    rev_y_smooth = y_smooth[::-1]\n",
    "\n",
    "    index = -1\n",
    "    mindist = 1000000\n",
    "    for i in np.arange(1,len(y_smooth)):\n",
    "        dist = abs(rev_y_smooth[i] - y_smooth[i])\n",
    "        if(dist < mindist):        \n",
    "            mindist = dist\n",
    "            index = i\n",
    "\n",
    "    ynew = np.zeros(shape=(len(y_smooth)))\n",
    "    for i in np.arange(0,index+1):\n",
    "        ynew[i] = rev_y_smooth[i] - 2*abs(rev_y_smooth[i] - y_smooth[index])\n",
    "\n",
    "    for i in np.arange(1,index+2):\n",
    "        ynew[i+index] = rev_y_smooth[i+index] + 2*abs(rev_y_smooth[i+index] - y_smooth[index])   \n",
    "\n",
    "    p2.line(x_smooth,y_smooth, color=\"gray\",line_alpha=0.5, line_width=6, legend=\"Contig %GC distribution\")\n",
    "    p2.line(x_smooth,ynew, color=\"red\",line_alpha=0.5,line_width=6, legend=\"Contig %GC distribution assymetry\")\n",
    "\n",
    "    p2.circle(x, y,\n",
    "                line_color=\"black\", fill_color=\"gray\", fill_alpha=1, size=2, legend=\"Contig %GC values\")\n",
    "\n",
    "    p2.xaxis.axis_label=\"Fraction contigs < %GC\"\n",
    "    p2.yaxis.axis_label=\"%GC\"\n",
    "    p2.legend.orientation = \"top_left\"\n",
    "    p2.xaxis.bounds = (0,1)\n",
    "    \n",
    "    \n",
    "    ###\n",
    "    xscale3 = max(contig_len_values) - min(contig_len_values)\n",
    "    radius_scale3 = xscale3/global_scale\n",
    "    p3 = figure(title=\"Contig %GC by contig length\",background_fill=\"#FFF5E0\",plot_width=400, plot_height=400,title_text_font_size='14pt')\n",
    "    p3.scatter(contig_len_values, contig_gc_values, marker=\"circle\",radius=log10_feature_counts*radius_scale3,\n",
    "                line_color=\"#6666ee\", fill_color=\"#ee6666\", fill_alpha=0.3, size=7, legend=\"Number of genes\")\n",
    "    p3.xaxis.axis_label=\"Contig length (bp)\"\n",
    "    p3.yaxis.axis_label=\"%GC\"\n",
    "    \n",
    "    global_scale3 = xscale3/radius_scale3\n",
    "    \n",
    "    \n",
    "    ###\n",
    "    xscale4 = max(log10_contig_len_values) - min(log10_contig_len_values)    \n",
    "    radius_scale4 = xscale4/global_scale\n",
    "    p4 = figure(title=\"Contig %GC by log10 contig length\",background_fill=\"#FFF5E0\",plot_width=400, plot_height=400,title_text_font_size='14pt') \n",
    "    p4.scatter(log10_contig_len_values, contig_gc_values, marker=\"circle\", radius=log10_feature_counts*radius_scale4,\n",
    "                line_color=\"#6666ee\", fill_color=\"#ee6666\", fill_alpha=0.3, size=7, legend=\"Number of genes\")\n",
    "    p4.xaxis.axis_label=\"Log10 Contig length (bp)\"\n",
    "    p4.yaxis.axis_label=\"%GC\"\n",
    "\n",
    "    global_scale4 = xscale4/radius_scale4\n",
    "\n",
    "    \n",
    "    \n",
    "    #global_scale910 = 10000.0\n",
    "    ### ###\n",
    "    xscale9 = max(contig_len_values) - min(contig_len_values)\n",
    "    radius_scale9 = xscale9/global_scale\n",
    "    p9 = figure(title=\"Feature count by contig length\",background_fill=\"#FFF5E0\",plot_width=400, plot_height=400,title_text_font_size='14pt')\n",
    "    p9.scatter(contig_len_values, log10_feature_counts, marker=\"circle\",radius=np.asarray(log10_contig_gc_values)*radius_scale9,\n",
    "                line_color=\"#6666ee\", fill_color=\"#ee6666\", fill_alpha=0.3, size=7, legend=\"%GC\")\n",
    "    p9.xaxis.axis_label=\"Contig length (bp)\"\n",
    "    p9.yaxis.axis_label=\"Log10 Number of genes\"\n",
    "    p9.legend.orientation = \"bottom_right\"\n",
    "    \n",
    "    global_scale9 = xscale9/radius_scale9\n",
    "    \n",
    "    \n",
    "    ###\n",
    "    xscale10 = max(log10_contig_len_values) - min(log10_contig_len_values)    \n",
    "    radius_scale10 = xscale10/global_scale\n",
    "    p10 = figure(title=\"Feature count by log10 contig length\",background_fill=\"#FFF5E0\",plot_width=400, plot_height=400,title_text_font_size='14pt') \n",
    "    p10.scatter(log10_contig_len_values, log10_feature_counts, marker=\"circle\", radius=np.asarray(log10_contig_gc_values)*radius_scale10,\n",
    "                line_color=\"#6666ee\", fill_color=\"#ee6666\", fill_alpha=0.3, size=7, legend=\"%GC\")\n",
    "    p10.xaxis.axis_label=\"Log10 Contig length (bp)\"\n",
    "    p10.yaxis.axis_label=\"Log10 Number of genes\"\n",
    "    p10.legend.orientation = \"bottom_right\"\n",
    "    \n",
    "    global_scale10 = xscale10/radius_scale10\n",
    "    \n",
    "    print xscale1, xscale3, xscale4, xscale9, xscale10\n",
    "    print global_scale1,global_scale3, global_scale4, global_scale9, global_scale10\n",
    "    \n",
    "       \n",
    "    ###\n",
    "    import scipy.special, scipy.stats\n",
    "    \n",
    "    mu5 = np.mean(contig_gc_values)\n",
    "    mu5print = \"{:3.1f}\".format(mu5)\n",
    "    sigma5 = np.std(contig_gc_values)\n",
    "    sigma5print = \"{:3.1f}\".format(sigma5)\n",
    "    skew5 = scipy.stats.skew(contig_gc_values)\n",
    "    skew5print = \"{:3.1f}\".format(skew5)\n",
    "    #print(skew5print)  \n",
    "    \n",
    "    p5 = figure(title=\"Contig %GC (μ={mu5print}, σ={sigma5print}, sk={skew5print})\".format(**vars()),tools=\"save\",\n",
    "       background_fill=\"#FFF5E0\",plot_width=400, plot_height=400,title_text_font_size='14pt')\n",
    "\n",
    "    #measured = np.random.normal(mu, sigma, 1000)\n",
    "    hist5, edges5 = np.histogram(contig_gc_values, density=True, bins=50)\n",
    "\n",
    "    x = np.linspace(min(contig_gc_values), max(contig_gc_values), 100)\n",
    "    pdf = 1/(sigma5 * np.sqrt(2*np.pi)) * np.exp(-(x-mu5)**2 / (2*sigma5**2))\n",
    "    cdf = (1+scipy.special.erf((x-mu5)/np.sqrt(2*sigma5**2)))/2\n",
    "\n",
    "    p5.quad(top=hist5, bottom=0, left=edges5[:-1], right=edges5[1:],\n",
    "         fill_color=\"#59D659\", line_color=\"#59D659\")\n",
    "    p5.line(x, pdf, line_color=\"#D17519\", line_width=4, alpha=0.7, legend=\"Normal PDF estimate\")\n",
    "    p5.line(x, cdf, line_color=\"#4D504D\", line_width=2, alpha=0.7, legend=\"CDF\")\n",
    "\n",
    "    p5.legend.orientation = \"top_left\"\n",
    "    p5.xaxis.axis_label = 'Contig %GC'\n",
    "    p5.yaxis.axis_label = 'Pr(Contig %GC)'\n",
    "\n",
    "    \n",
    "    ###\n",
    "    contig_lengths_log10 = np.log10(contig_len_values)\n",
    "    mu6 = np.mean(contig_lengths_log10)\n",
    "    mu6print = \"{:2.0f}\".format(mu6)\n",
    "    sigma6 = np.std(contig_lengths_log10)\n",
    "    sigma6print = \"{:2.0f}\".format(sigma6)\n",
    "    skew6 = scipy.stats.skew(contig_lengths_log10)\n",
    "    skew6print = \"{:3.1f}\".format(skew6)\n",
    "    \n",
    "    p6 = figure(title=\"Log10 Contig lengths (μ={mu6print}, σ={sigma6print}, sk={skew6print})\".format(**vars()),tools=\"save\",\n",
    "       background_fill=\"#FFF5E0\",plot_width=400, plot_height=400,title_text_font_size='14pt')\n",
    "\n",
    "    #measured = np.random.normal(mu, sigma, 1000)\n",
    "    hist6, edges6 = np.histogram(contig_lengths_log10, density=True, bins=50)\n",
    "    #print hist6\n",
    "    #print contig_lengths_log10\n",
    "    \n",
    "    x = np.linspace(min(contig_lengths_log10), max(contig_lengths_log10), 100)\n",
    "    pdf = 1/(sigma6 * np.sqrt(2*np.pi)) * np.exp(-(x-mu6)**2 / (2*sigma6**2))\n",
    "    cdf = (1+scipy.special.erf((x-mu6)/np.sqrt(2*sigma6**2)))/2\n",
    "\n",
    "    p6.quad(top=hist6, bottom=0, left=edges6[:-1], right=edges6[1:],\n",
    "         fill_color=\"#59D659\", line_color=\"#59D659\")\n",
    "    p6.line(x, pdf, line_color=\"#D17519\", line_width=4, alpha=0.7, legend=\"Normal PDF estimate\")\n",
    "    p6.line(x, cdf, line_color=\"#4D504D\", line_width=2, alpha=0.7, legend=\"CDF\")\n",
    "\n",
    "    p6.legend.orientation = \"top_left\"\n",
    "    p6.xaxis.axis_label = 'Log10 Contig lengths (bp)'\n",
    "    p6.yaxis.axis_label = 'Pr(Length)'\n",
    "    \n",
    "    \n",
    "    ###\n",
    "    \n",
    "    mu7 = np.mean(log10_feature_counts)\n",
    "    mu7print = \"{:2.0f}\".format(mu7)\n",
    "    sigma7 = np.std(log10_feature_counts)\n",
    "    sigma7print = \"{:2.0f}\".format(sigma7)\n",
    "    skew7 = scipy.stats.skew(log10_feature_counts)\n",
    "    skew7print = \"{:3.1f}\".format(skew7)\n",
    "    \n",
    "    p7 = figure(title=\"Log10 number of genes (μ={mu6print}, σ={sigma6print}, sk={skew6print})\".format(**vars()),tools=\"save\",\n",
    "       background_fill=\"#FFF5E0\",plot_width=400, plot_height=400,title_text_font_size='14pt')\n",
    "\n",
    "    #measured = np.random.normal(mu, sigma, 1000)\n",
    "    hist7, edges7 = np.histogram(log10_feature_counts, density=True, bins=50)\n",
    "    #print hist6\n",
    "    #print contig_lengths_log10\n",
    "    \n",
    "    x = np.linspace(min(log10_feature_counts), max(log10_feature_counts), 100)\n",
    "    pdf = 1/(sigma7 * np.sqrt(2*np.pi)) * np.exp(-(x-mu7)**2 / (2*sigma7**2))\n",
    "    cdf = (1+scipy.special.erf((x-mu7)/np.sqrt(2*sigma7**2)))/2\n",
    "\n",
    "    p7.quad(top=hist7, bottom=0, left=edges7[:-1], right=edges7[1:],\n",
    "         fill_color=\"#59D659\", line_color=\"#59D659\")\n",
    "    p7.line(x, pdf, line_color=\"#D17519\", line_width=4, alpha=0.7, legend=\"Normal PDF estimate\")\n",
    "    p7.line(x, cdf, line_color=\"#4D504D\", line_width=2, alpha=0.7, legend=\"CDF\")\n",
    "\n",
    "    p7.legend.orientation = \"top_left\"\n",
    "    p7.xaxis.axis_label = 'Log10 number of genes'\n",
    "    p7.yaxis.axis_label = 'Pr(Length)'\n",
    "    \n",
    "    ###  \n",
    "    import operator\n",
    "\n",
    "    length_dict = dict()\n",
    "    count = 0\n",
    "    for c in contig_ids:\n",
    "        length_dict[c] = contig_len_values[count]\n",
    "        count += 1\n",
    "\n",
    "    ###code to sort along with contig ids\n",
    "    #sorted_length = sorted(length_dict.items(), key=operator.itemgetter(1))\n",
    "\n",
    "    sorted_len_ar = np.asarray(sorted(length_dict.values()))[::-1]\n",
    "    sum = np.sum(sorted_len_ar)\n",
    "    cumsum = np.cumsum(sorted_len_ar)\n",
    "    cumsum_norm = 100.0 * cumsum.astype('f') / (sum + 0.0)\n",
    "\n",
    "    xs = np.zeros(2 * len(cumsum_norm)-2)\n",
    "    ys = np.zeros(2 * len(cumsum_norm)-2)\n",
    "    xs[0] = 0\n",
    "    xs[1] = cumsum_norm[0]\n",
    "    ys[0] = sorted_len_ar[0]\n",
    "    ys[1] = sorted_len_ar[0]\n",
    "\n",
    "    for i in  range(1,len(cumsum_norm)-1,1):\n",
    "        xs[2*i] = cumsum_norm[i-1]\n",
    "        xs[2*i+1] = cumsum_norm[i]\n",
    "        ys[2*i] = sorted_len_ar[i]\n",
    "        ys[2*i+1] = sorted_len_ar[i]\n",
    "\n",
    "    maxy =  max(ys)\n",
    "    p8 = figure(title=\"Nx plot for contig length\",background_fill=\"#FFF5E0\",plot_width=400, plot_height=400,title_text_font_size='14pt')\n",
    "    p8.line(xs, ys, line_color=\"#6666ee\", line_width=2, alpha=0.7,legend=\"Nx\")\n",
    "    p8.line([50, 50], [0,maxy], line_color=\"red\", line_width=2, alpha=0.7,legend=\"N50\")\n",
    "    ###the below line prints NaN, puzzling\n",
    "    #p8.text(50, maxy, \"N50 = \")\n",
    "    p8.xaxis.axis_label=\"Nx\"\n",
    "    p8.yaxis.axis_label=\"Contig length\"\n",
    "\n",
    "    \n",
    "    masterp = gridplot([[p1, p2], [p3, p4], [p9, p10], [p5, p7],[p6, p8]])\n",
    "\n",
    "    show(masterp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def run_genome_quality(genomeref='3847_EnsembleRelease30'):\n",
    "    \n",
    "    #token = None\n",
    "    s = \"\"\n",
    "    #annotations = {genome: GenomeAnnotationAPI(services, token, s.join((\"ReferenceEnsemblPlantGenomeAnnotations/\",genome)))\n",
    "    #               #\"kb|g.166819\": GenomeAnnotationAPI(services, token, \"PrototypeReferenceGenomes/kb|g.166819\"),\n",
    "    #               #\"kb|g.3899\": GenomeAnnotationAPI(services, token, \"PrototypeReferenceGenomes/kb|g.3899\")\n",
    "    #              }\n",
    "    \n",
    "    annoapi = GenomeAnnotationClientAPI(genomeannotationurl, token, s.join((\"ReferenceEnsemblPlantGenomeAnnotations/\",genomeref)))\n",
    "    annotations = {genomeref: annoapi} \n",
    "    start_time = datetime.datetime.utcnow()\n",
    "    for n in annotations:\n",
    "        print n\n",
    "        data = dict()\n",
    "        overview = get_genome_summary(annotations[n])\n",
    "\n",
    "        contig_ids = sorted(overview[\"assembly\"][\"contig_length\"].keys(), \n",
    "                            cmp=lambda x,y: cmp(x.rsplit(\".\",1)[1],y.rsplit(\".\",1)[1]))\n",
    "\n",
    "        contig_len_values = [overview[\"assembly\"][\"contig_length\"][c] for c in contig_ids]\n",
    "        contig_gc_values = [100.0 * overview[\"assembly\"][\"contig_gc_content\"][c] for c in contig_ids]\n",
    "\n",
    "\n",
    "        count =0\n",
    "        start=1\n",
    "        stop=100000000\n",
    "        #for c in contig_ids:\n",
    "        #    print c\n",
    "\n",
    "        regions = []\n",
    "        for c in contig_ids:\n",
    "            contig_add = {\"contig_id\": c, \"start\": start, \"length\": stop-start, \"strand\": \"?\"}\n",
    "            regions.append(contig_add)\n",
    "\n",
    "\n",
    "            \n",
    "        feat_data = annotations[n].get_feature_ids(filters={\"region_list\":regions})\n",
    "\n",
    "        print feat_data\n",
    "        \n",
    "        feature_counts = []    \n",
    "        for c in contig_ids:\n",
    "            num = len(feat_data['region'][c])\n",
    "            if(num == 0):\n",
    "                num = 0.1\n",
    "            feature_counts.append(num)\n",
    "\n",
    "        plot_genome_quality(contig_gc_values, contig_len_values,feature_counts)\n",
    "\n",
    "        print \"WARNING breaking after first genome\"\n",
    "        break;\n",
    "    end_time = datetime.datetime.utcnow()\n",
    "    print \"Total time : {0}\".format(end_time - start_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3847_EnsembleRelease30\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-68-e577da2659cb>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mrun_genome_quality\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'3847_EnsembleRelease30'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;31m#ReferenceEnsemblPlantGenomeAnnotations\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-67-30e14ec159f3>\u001b[0m in \u001b[0;36mrun_genome_quality\u001b[0;34m(genomeref)\u001b[0m\n\u001b[1;32m     14\u001b[0m         \u001b[0;32mprint\u001b[0m \u001b[0mn\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m         \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m         \u001b[0moverview\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_genome_summary\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mannotations\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m         contig_ids = sorted(overview[\"assembly\"][\"contig_length\"].keys(), \n",
      "\u001b[0;32m<ipython-input-53-dd8e0fc503ec>\u001b[0m in \u001b[0;36mget_genome_summary\u001b[0;34m(genome_annotation)\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0;31m#    raise TypeError(\"{0} is not a recognized GenomeAnnotation type.\".format(type(genome_annotation)))\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m     \u001b[0mtaxon\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgenome_annotation\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_taxon\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m     \u001b[0;32mprint\u001b[0m \u001b[0;34m\"genome_annotation.get_taxon \"\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mtaxon\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m     \u001b[0massemblyref\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgenome_annotation\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_assembly\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/doekbase/data_api/util.pyc\u001b[0m in \u001b[0;36mmethod_wrapper\u001b[0;34m(self, *args, **kwds)\u001b[0m\n\u001b[1;32m     98\u001b[0m         \u001b[0;32mdef\u001b[0m \u001b[0mmethod_wrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     99\u001b[0m             \u001b[0mt0\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlog_start\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlogger\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfull_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlog_level\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkvp\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkvp_str\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 100\u001b[0;31m             \u001b[0mreturnval\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    101\u001b[0m             \u001b[0mlog_end\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlogger\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mt0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfull_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlog_level\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkvp\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkvp_str\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    102\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mreturnval\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/doekbase/data_api/annotation/genome_annotation/api.pyc\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1746\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1747\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1748\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1749\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mttypes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mAttributeException\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1750\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mAttributeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmessage\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/doekbase/data_api/annotation/genome_annotation/api.pyc\u001b[0m in \u001b[0;36mget_taxon\u001b[0;34m(self, ref_only)\u001b[0m\n\u001b[1;32m   1776\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mclient_method\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1777\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mget_taxon\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mref_only\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1778\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclient\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_taxon\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_token\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mref\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1779\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1780\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mlogged\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_ga_log\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/doekbase/data_api/annotation/genome_annotation/service/thrift_client.pyc\u001b[0m in \u001b[0;36mget_taxon\u001b[0;34m(self, token, ref)\u001b[0m\n\u001b[1;32m    285\u001b[0m      \u001b[0;34m-\u001b[0m \u001b[0mref\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    286\u001b[0m     \"\"\"\n\u001b[0;32m--> 287\u001b[0;31m     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend_get_taxon\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtoken\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mref\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    288\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_get_taxon\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    289\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/doekbase/data_api/annotation/genome_annotation/service/thrift_client.pyc\u001b[0m in \u001b[0;36msend_get_taxon\u001b[0;34m(self, token, ref)\u001b[0m\n\u001b[1;32m    295\u001b[0m     \u001b[0margs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_oprot\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    296\u001b[0m     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_oprot\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwriteMessageEnd\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 297\u001b[0;31m     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_oprot\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtrans\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflush\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    298\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    299\u001b[0m   \u001b[0;32mdef\u001b[0m \u001b[0mrecv_get_taxon\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/thrift/transport/THttpClient.pyc\u001b[0m in \u001b[0;36m_f\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    104\u001b[0m       \u001b[0morig_timeout\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msocket\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetdefaulttimeout\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    105\u001b[0m       \u001b[0msocket\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msetdefaulttimeout\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__timeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 106\u001b[0;31m       \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    107\u001b[0m       \u001b[0msocket\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msetdefaulttimeout\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0morig_timeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    108\u001b[0m       \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/marcin/Documents/KBase/modules/data_api/venv/lib/python2.7/site-packages/thrift/transport/THttpClient.pyc\u001b[0m in \u001b[0;36mflush\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    143\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    144\u001b[0m     \u001b[0;31m# Get reply to flush the request\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 145\u001b[0;31m     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmessage\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mheaders\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__http\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetreply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    146\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    147\u001b[0m   \u001b[0;31m# Decorate if we know how to timeout\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/httplib.pyc\u001b[0m in \u001b[0;36mgetreply\u001b[0;34m(self, buffering)\u001b[0m\n\u001b[1;32m   1205\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1206\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mbuffering\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1207\u001b[0;31m                 \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_conn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetresponse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1208\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1209\u001b[0m                 \u001b[0;31m#only add this keyword if non-default for compatibility\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/httplib.pyc\u001b[0m in \u001b[0;36mgetresponse\u001b[0;34m(self, buffering)\u001b[0m\n\u001b[1;32m   1130\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1131\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1132\u001b[0;31m             \u001b[0mresponse\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbegin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1133\u001b[0m             \u001b[0;32massert\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwill_close\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0m_UNKNOWN\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1134\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__state\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_CS_IDLE\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/httplib.pyc\u001b[0m in \u001b[0;36mbegin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    451\u001b[0m         \u001b[0;31m# read until we get a non-100 response\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    452\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 453\u001b[0;31m             \u001b[0mversion\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstatus\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreason\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_read_status\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    454\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mstatus\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mCONTINUE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    455\u001b[0m                 \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/httplib.pyc\u001b[0m in \u001b[0;36m_read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    407\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_read_status\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    408\u001b[0m         \u001b[0;31m# Initialize with Simple-Response defaults\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 409\u001b[0;31m         \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_MAXLINE\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    410\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0m_MAXLINE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    411\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mLineTooLong\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"header line\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/socket.pyc\u001b[0m in \u001b[0;36mreadline\u001b[0;34m(self, size)\u001b[0m\n\u001b[1;32m    478\u001b[0m             \u001b[0;32mwhile\u001b[0m \u001b[0mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    479\u001b[0m                 \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 480\u001b[0;31m                     \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_rbufsize\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    481\u001b[0m                 \u001b[0;32mexcept\u001b[0m \u001b[0merror\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    482\u001b[0m                     \u001b[0;32mif\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mEINTR\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ssl.pyc\u001b[0m in \u001b[0;36mrecv\u001b[0;34m(self, buflen, flags)\u001b[0m\n\u001b[1;32m    732\u001b[0m                     \u001b[0;34m\"non-zero flags not allowed in calls to recv() on %s\"\u001b[0m \u001b[0;34m%\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    733\u001b[0m                     self.__class__)\n\u001b[0;32m--> 734\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbuflen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    735\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    736\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbuflen\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mflags\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ssl.pyc\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, len, buffer)\u001b[0m\n\u001b[1;32m    619\u001b[0m                 \u001b[0mv\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sslobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbuffer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    620\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 621\u001b[0;31m                 \u001b[0mv\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sslobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;36m1024\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    622\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    623\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mSSLError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "run_genome_quality('3847_EnsembleRelease30')#ReferenceEnsemblPlantGenomeAnnotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
